### PR TITLE
Step Detection From Head and Hands Tracking

### PR DESCRIPTION
**Pull request by amantley**
This pr heuristically derived thresholds, to detect steps from head and hand tracking data.
Also in this pr, two properties were added to MyAvatar.h that allow setting the filter length for the average head yaw, and the threshold angle that leads to rotational recentering avatar to match the head yaw.

The associated fogbugz ticket is https://highfidelity.manuscript.com/f/cases/5338/Better-Recentering-Leaning-Turning-vs-Stepping

### Test Plan:
1. Start interface and load the attached mirror.json file using create->load json
2. Side-step to the right. You should see your avatar take a step to the right.
3. Side-step to the left.  You should see your avatar take a step to the left
4. Lean your body to the right, you should see your avatar lean to the right.
5. Lean your body to the left, you should see your avatar lean to the left.
6. Bend forward and you should see your avatar bend forward
7. Lean back and you should see your avatar lean back
8. Step forward with either foot and your avatar should step forward with the right foot
9. Step back with either foot and your avatar will step back with the right foot.
10. Go up on your toes briefly and nothing should change about the above behaviours

The list of conditions for stepping that are included in MyAvatar.cpp are the following:

1. has the head moved off of the base of support of the feet?  withinBaseOfSupport()
2. is the angular velocity of the head(roll and pitch) less than the threshold? headAngularVelocity()
3. does the hand velocity direction match the head?  handDirectionMatchesHeadDirection()
4. is the angular velocity of the hands(roll and pitch) less than the threshold? handAngularVelocityBelowThreshold
5. is the head velocity over a threshold that distinquishes slow sway from a step?  headVelocityGreaterThanThreshold()
6. is the head level compared to it average orientation with the xz plane?  isHeadLevel()

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11077/HighFidelity-Beta-PR13528-f216316.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11077/HighFidelity-Beta-PR13528-f216316.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11077/HighFidelity-Beta-Interface-PR13528-f216316.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11108/HighFidelity-Beta-PR13528-4995b57.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11108/HighFidelity-Beta-PR13528-4995b57.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11108/HighFidelity-Beta-Interface-PR13528-4995b57.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11114/HighFidelity-Beta-PR13528-6460c9e.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11114/HighFidelity-Beta-PR13528-6460c9e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11114/HighFidelity-Beta-Interface-PR13528-6460c9e.exe

**Comment by amantley** [comment]
I got feedback from Philip about the problem of the failsafe happening too often.  This was caused mainly by the mode being raised when you go up on your toes, which caused stepping to not be detected.  I fixed this by keeping long term frequencies for all the heights.  This means that if you return to your heels the mode will now go back down with out a reset.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11144/HighFidelity-Beta-PR13528-0ebc7fc.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11144/HighFidelity-Beta-PR13528-0ebc7fc.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11144/HighFidelity-Beta-Interface-PR13528-0ebc7fc.exe

File: interface/src/avatar/MyAvatar.cpp [hunk 1]
```diff
@@ -3135,6 +3152,132 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+    if (head.isValid()) {
+        isWithinThreshold = (head.getTranslation().y - newMode) > DEFAULT_AVATAR_MODE_HEIGHT_STEPPING_THRESHOLD;
+    }
+    return isWithinThreshold;
+}
+
+float MyAvatar::computeStandingHeightMode(controller::Pose head) {
+    const float CENTIMETERS_PER_METER = 100.0f;
+    const float MODE_CORRECTION_FACTOR = 0.02f;
+    const int MAX_INT = 2147483647;
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: use std::numeric_limits to get the max int value.

File: interface/src/avatar/MyAvatar.h [hunk 2]
```diff
@@ -1506,6 +1523,12 @@ private slots:
     // cache head controller pose in sensor space
     glm::vec2 _headControllerFacing;  // facing vector in xz plane (sensor space)
     glm::vec2 _headControllerFacingMovingAverage { 0.0f, 0.0f };   // facing vector in xz plane (sensor space)
+    glm::quat _averageHeadRotation { 0.0f, 0.0f, 0.0f, 1.0f };
+
+    float _standingHeightMode { 0.0f };
+    bool _resetMode { true };
+    std::map<long int, long int> _heightFrequencyMap;
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.h]: I'm concerned with this being a map instead of something with fixed storage cost and cache locality.  (std::vector or static array)  Can we assume reasonable human height limits here instead?

File: interface/src/avatar/MyAvatar.cpp [hunk 3]
```diff
@@ -3135,6 +3152,132 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+    // init mode in meters to the current mode
+    float modeInMeters = getStandingHeightMode();
+    if (head.isValid()) {
+        float newReading = head.getTranslation().y;
+        int newReadingInCentimeters = glm::floor(newReading * CENTIMETERS_PER_METER);
+        _heightFrequencyMap[newReadingInCentimeters] += 1;
+        if (_heightFrequencyMap[newReadingInCentimeters] > _greatestFrequency) {
+            _greatestFrequency = _heightFrequencyMap[newReadingInCentimeters];
+            modeInMeters = ((float)newReadingInCentimeters) / CENTIMETERS_PER_METER;
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: If I'm reading this correctly entries in the _heightFrequencyMap are only incremented.  So if you somehow switch from sitting for 10 minutes to standing, it will take another 10 minutes for the mode to pick up the new standing height.

I'm thinking that it might be better to use a different approach.  Maybe keep a ringBuffer of the last N (500?) head.y values, in a vector.  Then build a histogram from that data, and pick the bin with the highest count.

File: interface/src/avatar/MyAvatar.cpp [hunk 4]
```diff
@@ -3135,6 +3152,132 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+    return modeInMeters;
+}
+
+static bool handDirectionMatchesHeadDirection(controller::Pose leftHand, controller::Pose rightHand, controller::Pose head) {
+    const float VELOCITY_EPSILON = 0.02f;
+    bool leftHandDirectionMatchesHead = true;
+    bool rightHandDirectionMatchesHead = true;
+    if (leftHand.isValid() && head.isValid()) {
+        leftHand.velocity.y = 0.0f;
+        float handDotHeadLeft = glm::dot(glm::normalize(leftHand.getVelocity()), glm::normalize(head.getVelocity()));
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: although rare, it is possible for the velocity to be zero.  We should check that case to avoid div by zero.

File: interface/src/avatar/MyAvatar.cpp [hunk 5]
```diff
@@ -3135,6 +3152,132 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
     return worldToSensorMat * avatarToWorldMat * avatarHipsMat;
 }
 
+static bool isInsideLine(glm::vec3 a, glm::vec3 b, glm::vec3 c) {
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: a, b, and c be const refs

File: interface/src/avatar/MyAvatar.cpp [hunk 6]
```diff
@@ -3135,6 +3152,132 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+        bool withinFrontBase = isInsideLine(userScale * frontLeft, userScale * frontRight, head.getTranslation());
+        bool withinBackBase = isInsideLine(userScale * backRight, userScale * backLeft, head.getTranslation());
+        bool withinLateralBase = (isInsideLine(userScale * frontRight, userScale * backRight, head.getTranslation()) &&
+            isInsideLine(userScale * backLeft, userScale * frontLeft, head.getTranslation()));
+        isWithinSupport = (withinFrontBase && withinBackBase && withinLateralBase);
+    }
+    return isWithinSupport;
+}
+
+static bool headAngularVelocityBelowThreshold(controller::Pose head) {
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: make head be a const ref

File: interface/src/avatar/MyAvatar.cpp [hunk 7]
```diff
@@ -3135,6 +3152,132 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+        xzPlaneAngularVelocity.x = head.getAngularVelocity().x;
+        xzPlaneAngularVelocity.z = head.getAngularVelocity().z;
+    }
+    float magnitudeAngularVelocity = glm::length(xzPlaneAngularVelocity);
+    bool isBelowThreshold = (magnitudeAngularVelocity < DEFAULT_AVATAR_HEAD_ANGULAR_VELOCITY_STEPPING_THRESHOLD);
+
+    return isBelowThreshold;
+}
+
+static bool isWithinThresholdHeightMode(controller::Pose head, float newMode) {
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: head should be a const ref

File: interface/src/avatar/MyAvatar.cpp [hunk 8]
```diff
@@ -3135,6 +3152,132 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+    bool isWithinThreshold = true;
+    if (head.isValid()) {
+        isWithinThreshold = (head.getTranslation().y - newMode) > DEFAULT_AVATAR_MODE_HEIGHT_STEPPING_THRESHOLD;
+    }
+    return isWithinThreshold;
+}
+
+float MyAvatar::computeStandingHeightMode(controller::Pose head) {
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: const ref

File: interface/src/avatar/MyAvatar.cpp [hunk 9]
```diff
@@ -3135,6 +3152,132 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+                    }
+                }
+                _greatestFrequency = 100;
+            }
+        }
+    }
+    return modeInMeters;
+}
+
+static bool handDirectionMatchesHeadDirection(controller::Pose leftHand, controller::Pose rightHand, controller::Pose head) {
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: pass arguments by const ref

File: interface/src/avatar/MyAvatar.cpp [hunk 10]
```diff
@@ -3135,6 +3152,132 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+    }
+    if (rightHand.isValid() && head.isValid()) {
+        rightHand.velocity.y = 0.0f;
+        float handDotHeadRight = glm::dot(glm::normalize(rightHand.getVelocity()), glm::normalize(head.getVelocity()));
+        rightHandDirectionMatchesHead = ((handDotHeadRight > DEFAULT_HANDS_VELOCITY_DIRECTION_STEPPING_THRESHOLD) && (glm::length(rightHand.getVelocity()) > VELOCITY_EPSILON));
+    }
+    return leftHandDirectionMatchesHead && rightHandDirectionMatchesHead;
+}
+
+static bool handAngularVelocityBelowThreshold(controller::Pose leftHand, controller::Pose rightHand) {
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: const ref

File: interface/src/avatar/MyAvatar.cpp [hunk 11]
```diff
@@ -3135,6 +3152,132 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+    }
+    if (rightHand.isValid()) {
+        rightHand.angularVelocity.y = 0.0f;
+        rightHandXZAngularVelocity = glm::length(rightHand.getAngularVelocity());
+    }
+    return ((leftHandXZAngularVelocity < DEFAULT_HANDS_ANGULAR_VELOCITY_STEPPING_THRESHOLD) &&
+        (rightHandXZAngularVelocity < DEFAULT_HANDS_ANGULAR_VELOCITY_STEPPING_THRESHOLD));
+}
+
+static bool headVelocityGreaterThanThreshold(controller::Pose head) {
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: const ref

File: interface/src/avatar/MyAvatar.cpp [hunk 12]
```diff
@@ -3135,6 +3152,132 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+    float headVelocityMagnitude = 0.0f;
+    if (head.isValid()) {
+        headVelocityMagnitude = glm::length(head.getVelocity());
+    }
+    return headVelocityMagnitude > DEFAULT_HEAD_VELOCITY_STEPPING_THRESHOLD;
+}
+
+glm::quat MyAvatar::computeAverageHeadRotation(controller::Pose head) {
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: const ref argument

File: interface/src/avatar/MyAvatar.cpp [hunk 13]
```diff
@@ -3135,6 +3152,132 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+    const float AVERAGING_RATE = 0.03f;
+    return slerp(_averageHeadRotation, head.getRotation(), AVERAGING_RATE);
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: is this safe for quats of different polarity?

File: interface/src/avatar/MyAvatar.h [hunk 14]
```diff
@@ -1506,6 +1523,12 @@ private slots:
     // cache head controller pose in sensor space
     glm::vec2 _headControllerFacing;  // facing vector in xz plane (sensor space)
     glm::vec2 _headControllerFacingMovingAverage { 0.0f, 0.0f };   // facing vector in xz plane (sensor space)
+    glm::quat _averageHeadRotation { 0.0f, 0.0f, 0.0f, 1.0f };
+
+    float _standingHeightMode { 0.0f };
+    bool _resetMode { true };
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.h]: Is this hooked up to anything? I don't see any code that references this or getResetMode(), I assume it's supposed to clear the _heightFrequencyMap?

File: interface/src/avatar/MyAvatar.cpp [hunk 15]
```diff
@@ -3355,15 +3528,15 @@ void MyAvatar::FollowHelper::prePhysicsUpdate(MyAvatar& myAvatar, const glm::mat
             activate(Vertical);
         }
     } else {
-        if (!isActive(Rotation) && getForceActivateRotation()) {
+        if (!isActive(Rotation) && (getForceActivateRotation() || shouldActivateRotation(myAvatar, desiredBodyMatrix, currentBodyMatrix) || hasDriveInput)) {
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: Why did you add this here, it's important to NOT ever activate recentering if hmdLeanRecenterEnabled is false.
Some of our users in Vive Pucks are starting to set hmdLeanRecenterEnabled to false so they can drive without recentering.

File: interface/src/avatar/MyAvatar.h [hunk 16]
```diff
@@ -1584,7 +1608,7 @@ private slots:
     mutable std::mutex _controllerPoseMapMutex;
 
     bool _centerOfGravityModelEnabled { true };
-    bool _hmdLeanRecenterEnabled { true };
+    bool _hmdLeanRecenterEnabled { false };
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.h]: This needs to be enabled by default.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11186/HighFidelity-Beta-PR13528-defd6f8.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11186/HighFidelity-Beta-PR13528-defd6f8.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11186/HighFidelity-Beta-Interface-PR13528-defd6f8.exe

File: interface/src/avatar/MyAvatar.cpp [hunk 17]
```diff
@@ -2172,6 +2179,19 @@ void MyAvatar::setHasAudioEnabledFaceMovement(bool hasAudioEnabledFaceMovement)
     _headData->setHasAudioEnabledFaceMovement(hasAudioEnabledFaceMovement);
 }
 
+void MyAvatar::setRotationRecenterFilterLength(float length) {
+    const float MINIMUM_ROTATION_RECENTER_FILTER_LENGTH = 0.01f;
+    if (length > MINIMUM_ROTATION_RECENTER_FILTER_LENGTH) {
+        _rotationRecenterFilterLength = length;
+    } else {
+        _rotationRecenterFilterLength = 0.01f;
+    }
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: This can just be:
   
    _rotationRecenterfilterLength = std::max(MINIMUM_ROTATION_RECENTER_FILTER_LENGTH, length);

File: interface/src/avatar/MyAvatar.h [hunk 18]
```diff
@@ -880,6 +883,12 @@ class MyAvatar : public Avatar {
     virtual void rebuildCollisionShape() override;
 
     const glm::vec2& getHeadControllerFacingMovingAverage() const { return _headControllerFacingMovingAverage; }
+    const float getStandingHeightMode() const { return _standingHeightMode; }
+    void setStandingHeightMode(float newMode) { _standingHeightMode = newMode; }
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.h]: So these don't actually get and set the mode but the last set standing height.  Can we just rename the get/set and member var to getCurrentStandingHeight()?

File: interface/src/avatar/MyAvatar.cpp [hunk 19]
```diff
@@ -414,14 +418,17 @@ void MyAvatar::reset(bool andRecenter, bool andReload, bool andHead) {
...
+    const float HMD_FACING_TIMESCALE = getRotationRecenterFilterLength();
+    
     float tau = deltaTime / HMD_FACING_TIMESCALE;
     _headControllerFacingMovingAverage = lerp(_headControllerFacingMovingAverage, _headControllerFacing, tau);
 
     if (_smoothOrientationTimer < SMOOTH_TIME_ORIENTATION) {
         _rotationChanged = usecTimestampNow();
         _smoothOrientationTimer += deltaTime;
     }
+    setStandingHeightMode(computeStandingHeightMode(getControllerPoseInAvatarFrame(controller::Action::HEAD)));
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: why not just call _recentModeReadings.insert() here?

File: interface/src/avatar/MyAvatar.h [hunk 20]
```diff
@@ -1506,6 +1524,11 @@ private slots:
     // cache head controller pose in sensor space
     glm::vec2 _headControllerFacing;  // facing vector in xz plane (sensor space)
     glm::vec2 _headControllerFacingMovingAverage { 0.0f, 0.0f };   // facing vector in xz plane (sensor space)
+    glm::quat _averageHeadRotation { 0.0f, 0.0f, 0.0f, 1.0f };
+
+    float _standingHeightMode { 0.0f };
+    bool _resetMode { true };
+    RingBufferHistory<quint64> _recentModeReadings;
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.h]: why a 64 bit int?  can this just be int?

File: interface/src/avatar/MyAvatar.cpp [hunk 21]
```diff
@@ -3135,6 +3155,148 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+        int newReadingInCentimeters = glm::floor(newReading * CENTIMETERS_PER_METER);
+        _recentModeReadings.insert(newReadingInCentimeters);
+        RingBufferHistory<quint64>::Iterator recentModeReadingsIterator = _recentModeReadings.begin();
+        RingBufferHistory<quint64>::Iterator end = _recentModeReadings.end();
+        std::map<int, int> freq;
+        do {
+            ++recentModeReadingsIterator;
+            freq[*recentModeReadingsIterator] += 1;
+            if ((freq[*recentModeReadingsIterator] > greatestFrequency) ||
+                (freq[*recentModeReadingsIterator] == MODE_READINGS_RING_BUFFER_SIZE)) {
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: I'm not sure why the freq[*recentModeReadingsIterator] == MODE_READINGS_RING_BUFFER_SIZE case is necessary. Can you explain?

File: interface/src/avatar/MyAvatar.cpp [hunk 22]
```diff
@@ -3135,6 +3155,148 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+    // init mode in meters to the current mode
+    float modeInMeters = getStandingHeightMode();
+    if (head.isValid()) {
+        float newReading = head.getTranslation().y;
+        int newReadingInCentimeters = glm::floor(newReading * CENTIMETERS_PER_METER);
+        _recentModeReadings.insert(newReadingInCentimeters);
+        RingBufferHistory<quint64>::Iterator recentModeReadingsIterator = _recentModeReadings.begin();
+        RingBufferHistory<quint64>::Iterator end = _recentModeReadings.end();
+        std::map<int, int> freq;
+        do {
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: Do loops are not commonly used and might be confusing for some programmers, I don't feel that a do loop is strictly necessary here.  What do you think?

File: interface/src/avatar/MyAvatar.cpp [hunk 23]
```diff
@@ -3135,6 +3155,148 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+            }
+        } while (recentModeReadingsIterator != end);
+
+        modeInMeters = ((float)mode) / CENTIMETERS_PER_METER;
+        if (!(modeInMeters > getStandingHeightMode())) {
+            // if not greater check for a reset
+            if (getResetMode() && qApp->isHMDMode()) {
+                setResetMode(false);
+                float resetModeInCentimeters = glm::floor((newReading - MODE_CORRECTION_FACTOR)*CENTIMETERS_PER_METER);
+                modeInMeters = (resetModeInCentimeters / CENTIMETERS_PER_METER);
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: shouldn't we clear the _recentModeReadings?

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11199/HighFidelity-Beta-PR13528-a2f8e02.apk

File: interface/src/avatar/MyAvatar.cpp [hunk 24]
```diff
@@ -3135,6 +3156,148 @@ glm::mat4 MyAvatar::deriveBodyUsingCgModel() const {
...
+        for(auto recentModeReadingsIterator = _recentModeReadings.begin(); recentModeReadingsIterator != _recentModeReadings.end(); ++recentModeReadingsIterator) {
+            freq[*recentModeReadingsIterator] += 1;
+            if (freq[*recentModeReadingsIterator] > greatestFrequency) {
+                greatestFrequency = freq[*recentModeReadingsIterator];
+                mode = *recentModeReadingsIterator;
+            }
+        }
+
+        modeInMeters = ((float)mode) / CENTIMETERS_PER_METER;
+        if (!(modeInMeters > getCurrentStandingHeight())) {
```
Comment (hyperlogic) [inline interface/src/avatar/MyAvatar.cpp]: This is always going to be true.  What are you trying to detect here? if the user stands up on their tippy toes?

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11199/HighFidelity-Beta-PR13528-a2f8e02.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11199/HighFidelity-Beta-Interface-PR13528-a2f8e02.exe

**Comment by qliacoombs** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/11206/HighFidelity-Beta-PR13528-cb421fd.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11206/HighFidelity-Beta-PR13528-cb421fd.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/11206/HighFidelity-Beta-Interface-PR13528-cb421fd.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11206/HighFidelity-Beta-PR13528-cb421fd.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/11206/HighFidelity-Beta-Interface-PR13528-cb421fd.exe

**Comment by qliejahner** [comment]
Currently testing

**Comment by qliejahner** [comment]
PASS @conklin94122 

image
